gapis/api/vulkan/frame_loop.go

```diff
@@ -126,129 +139,237 @@ func newFrameLoop(ctx context.Context, c *capture.GraphicsCapture, numInitialCmd
 		imageChanged:   make(map[VkImage]bool),
 		imageDestroyed: make(map[VkImage]bool),
 		imageToBackup:  make(map[VkImage]VkImage),
-	}
 
-	f.loopStartCmd, f.loopEndCmd = f.getLoopStartAndEndCmd(ctx, Cmds)
-
-	return f
+		loopTerminated:      false,
+		lastObservedCommand: api.CmdNoID,
+	}
 }
 
-func (f *frameLoop) Transform(ctx context.Context, id api.CmdID, cmd api.Cmd, out transform.Writer) {
+func (f *frameLoop) Transform(ctx context.Context, cmdId api.CmdID, cmd api.Cmd, out transform.Writer) {
+
 	ctx = log.Enter(ctx, "FrameLoop Transform")
+	log.D(ctx, "FrameLoop: looping from %v to %v.", f.loopStartIdx, f.loopEndIdx)
+
+	// Lets capture and update the last observed frame from f. From this point on use the local lastObservedCommand variable.
+	lastObservedCommand := f.lastObservedCommand
+	f.lastObservedCommand = cmdId
 
-	if cmd == f.loopStartCmd {
-		log.D(ctx, "Loop: start loop at frame %v, id %v, cmd %v.", f.frameNum, id, cmd)
-		f.detectChangedResource(ctx, out.State())
+	if lastObservedCommand != api.CmdNoID && lastObservedCommand > api.CmdID.Real(cmdId) {
+		log.F(ctx, true, "FrameLoop: expected next observed command ID to be >= last observed command ID")
+	}
+
+	// Walk the frame count forwards if we just hit the end of one.
+	if _, ok := cmd.(*VkQueuePresentKHR); ok {
+		f.frameNum++
+	}
 
-		st := GetState(out.State())
-		sb := st.newStateBuilder(ctx, newTransformerOutput(out))
-		defer sb.ta.Dispose()
+	// Are we before the loop or just at the start of it?
+	if lastObservedCommand == api.CmdNoID {
+
+		// This is the start of the loop.
+		if api.CmdID.Real(cmdId) == f.loopStartIdx {
```

Comment:
If this the point where loop starts, then probably we need to backup the state here before further mutations.

Reply:
This is queuing up commands, and then only mutating them in detectChangedResource. That is where the backup is created.

Reply:
@AWoloszyn is this issue void?

Reply:
I think that is correct, from what I can tell this code is as expected. (Since nothing mutates)